[CI Failure]: Pip failure, pip not installing properly.

[Debugging Doc](https://docs.google.com/document/d/1vveidDIVcl-Bfh7gF4ZN_Ahk7cmb7q_9o7O-0tYGW94/edit?tab=t.0#heading=h.rpq25vez37fp)

### CI Test Type

All tests

### Which CI step failed?

Merge develop and setup up dependencies

### Stacktrace

```shell
Error: Unable to download artifact(s): Artifact not found for name: cached_build_files
        Please ensure that your artifact is not expired and the artifact was uploaded using a compatible version of toolkit/upload-artifact.
        For more information, visit the GitHub Artifacts FAQ: https://github.com/actions/toolkit/blob/main/packages/artifact/docs/faq.md
Run echo "Failed to download build files. Regenerating."
Failed to download build files. Regenerating.

Notice:  A new release of pip is available: 24.3.1 -> 25.1
Notice:  To update, run: pip install --upgrade pip
ERROR: Exception:
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 105, in _run_wrapper
    status = _inner_run()
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 96, in _inner_run
    return self.run(options, args)
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 67, in wrapper
    return func(self, options, args)
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/commands/install.py", line 363, in run
    resolver = self.make_resolver(
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 175, in make_resolver
    import pip._internal.resolution.resolvelib.resolver
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 8, in <module>
    from pip._vendor.resolvelib import BaseReporter, ResolutionImpossible
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/__init__.py", line 19, in <module>
    from .resolvers import (
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers/__init__.py", line 1, in <module>
    from ..structs import RequirementInformation
ImportError: cannot import name 'RequirementInformation' from 'pip._vendor.resolvelib.structs' (/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/structs.py)

Notice:  A new release of pip is available: 24.3.1 -> 25.1
Notice:  To update, run: pip install --upgrade pip
ERROR: Exception:
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 105, in _run_wrapper
    status = _inner_run()
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 96, in _inner_run
    return self.run(options, args)
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 67, in wrapper
    return func(self, options, args)
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/commands/install.py", line 363, in run
    resolver = self.make_resolver(
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 175, in make_resolver
    import pip._internal.resolution.resolvelib.resolver
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 8, in <module>
    from pip._vendor.resolvelib import BaseReporter, ResolutionImpossible
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/__init__.py", line 19, in <module>
    from .resolvers import (
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers/__init__.py", line 1, in <module>
    from ..structs import RequirementInformation
ImportError: cannot import name 'RequirementInformation' from 'pip._vendor.resolvelib.structs' (/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/structs.py)

Notice:  A new release of pip is available: 24.3.1 -> 25.1
Notice:  To update, run: pip install --upgrade pip
ERROR: Exception:
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 105, in _run_wrapper
    status = _inner_run()
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 96, in _inner_run
    return self.run(options, args)
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 67, in wrapper
    return func(self, options, args)
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/commands/install.py", line 363, in run
    resolver = self.make_resolver(
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 175, in make_resolver
    import pip._internal.resolution.resolvelib.resolver
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 8, in <module>
    from pip._vendor.resolvelib import BaseReporter, ResolutionImpossible
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/__init__.py", line 19, in <module>
    from .resolvers import (
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers/__init__.py", line 1, in <module>
    from ..structs import RequirementInformation
ImportError: cannot import name 'RequirementInformation' from 'pip._vendor.resolvelib.structs' (/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/structs.py)
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.20/x64/bin/pip-compile", line 5, in <module>
    from piptools.scripts.compile import cli
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/piptools/scripts/compile.py", line 25, in <module>
    from ..resolver import BacktrackingResolver, LegacyResolver
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/piptools/resolver.py", line 20, in <module>
    from pip._internal.resolution.resolvelib.resolver import Resolver
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 8, in <module>
    from pip._vendor.resolvelib import BaseReporter, ResolutionImpossible
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/__init__.py", line 19, in <module>
    from .resolvers import (
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers/__init__.py", line 1, in <module>
    from ..structs import RequirementInformation
ImportError: cannot import name 'RequirementInformation' from 'pip._vendor.resolvelib.structs' (/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/pip/_vendor/resolvelib/structs.py)
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/runner/work/oppia/oppia/scripts/build.py", line 42, in <module>
    from scripts import install_third_party_libs
  File "/home/runner/work/oppia/oppia/scripts/install_third_party_libs.py", line 46, in <module>
    install_python_dev_dependencies.main(['--assert_compiled'])
  File "/home/runner/work/oppia/oppia/scripts/install_python_dev_dependencies.py", line 154, in main
    not_compiled = compile_pip_requirements(
  File "/home/runner/work/oppia/oppia/scripts/install_python_dev_dependencies.py", line 134, in compile_pip_requirements
    subprocess.run(
  File "/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['pip-compile', '--no-emit-index-url', '--quiet', '--strip-extras', '--generate-hashes', 'requirements_dev.in', '--output-file', 'requirements_dev.txt']' returned non-zero exit status 1.
Error: Process completed with exit code 1.
```

### Screenshots (if e2e tests)

_No response_

### Occurrences

https://github.com/oppia/oppia/actions/runs/14731793635/job/41348109954

### Additional Information

_No response_

### Debugging document link

_No response_

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -36,6 +36,7 @@ runs:
         path: |
           /opt/hostedtoolcache/Python/3.9.20/x64
 
+    # TODO (#22567): Fix pip not working properly when fetched from cache.
     - name: Check pip installation and fix if needed
       run: |
         set -e
@@ -45,7 +46,8 @@ runs:
           python -m pip cache purge || true
           echo "Reinstalling pip and dev dependencies via install_python_dev_dependencies.py..."
           curl -sS https://bootstrap.pypa.io/get-pip.py -o get-pip.py
-          python get-pip.py
+          python -m pip uninstall pip -y
+          python get-pip.py pip==24.3.1
           python -m scripts.install_python_dev_dependencies
         else
           echo "pip is working correctly."

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -36,39 +36,22 @@ runs:
         path: |
           /opt/hostedtoolcache/Python/3.9.20/x64
 
-    - name: Validate and fix pip installation
-      id: validate_pip
+    - name: Check pip installation and fix if needed
       run: |
-        echo "Checking pip installation..."
-        # Try to run pip debug to check if pip is working
-        if ! python -m pip debug; then
-          echo "Pip appears to be broken, need to fix before proceeding"
-          # Remove pip completely
-          python -m pip uninstall -y pip setuptools
-
-          # Use get-pip.py to get a basic pip installation working
-          # In Install third-party Python dev dependencies step,
-          # correct version of pip will be installed
-          curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py
-          python get-pip.py
-
-          # Flag that we had to repair pip
-          echo "pip_repaired=true" >> $GITHUB_OUTPUT
+        set -e
+        echo "Checking if pip is properly installed..."
+        if ! python -m pip debug > /dev/null 2>&1; then
+          echo "pip is broken. Clearing pip cache..."
+          python -m pip cache purge || true
+          echo "Reinstalling pip and dev dependencies via install_python_dev_dependencies.py..."
+          python -m scripts.install_python_dev_dependencies
         else
-          echo "Pip appears to be working"
-          echo "pip_repaired=false" >> $GITHUB_OUTPUT
+          echo "pip is working correctly."
         fi
       shell: bash
 
-    - name: Clear pip cache if pip was repaired
-      if: steps.validate_pip.outputs.pip_repaired == 'true'
-      run: |
-        echo "Clearing pip cache due to repair"
-        rm -rf ~/.cache/pip
-      shell: bash
-
     - name: Install third-party Python dev dependencies
-      if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit == false || steps.validate_pip.outputs.pip_repaired == 'true'
+      if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit == false
       uses: nick-fields/retry@v3
       with:
         max_attempts: 3

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -44,6 +44,8 @@ runs:
           echo "pip is broken. Clearing pip cache..."
           python -m pip cache purge || true
           echo "Reinstalling pip and dev dependencies via install_python_dev_dependencies.py..."
+          curl -sS https://bootstrap.pypa.io/get-pip.py -o get-pip.py
+          python get-pip.py
           python -m scripts.install_python_dev_dependencies
         else
           echo "pip is working correctly."

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -45,9 +45,9 @@ runs:
           echo "pip is broken. Clearing pip cache..."
           python -m pip cache purge || true
           echo "Reinstalling pip and dev dependencies via install_python_dev_dependencies.py..."
-          curl -sS https://bootstrap.pypa.io/get-pip.py -o get-pip.py
+          curl -sS https://bootstrap.pypa.io/get-pip.py -o /tmp/get-pip.py
           python -m pip uninstall pip -y
-          python get-pip.py pip==24.3.1
+          python /tmp/get-pip.py pip==24.3.1
           python -m scripts.install_python_dev_dependencies
         else
           echo "pip is working correctly."

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -37,11 +37,14 @@ runs:
           /opt/hostedtoolcache/Python/3.9.20/x64
 
     # TODO (#22567): Fix pip not working properly when fetched from cache.
-    - name: Check cached pip installation and fix if needed
+    - name: Check pip installation and fix if needed
       run: |
         set -e
         echo "Checking if pip is properly installed..."
-        if ! python -m pip debug > /dev/null 2>&1; then
+        
+        # This shows pip-related environment and configuration info for debugging:
+        # https://pip.pypa.io/en/stable/cli/pip_debug/
+        if ! python -m pip debug; then
           echo "pip is broken. Clearing pip cache..."
           python -m pip cache purge || true
           echo "Reinstalling pip and dev dependencies via install_python_dev_dependencies.py..."

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -37,7 +37,7 @@ runs:
           /opt/hostedtoolcache/Python/3.9.20/x64
 
     - name: Validate and fix pip installation
-      id: validate_pip          
+      id: validate_pip
       run: |
         echo "Checking pip installation..."
         # Try to run pip debug to check if pip is working
@@ -58,6 +58,7 @@ runs:
           echo "Pip appears to be working"
           echo "pip_repaired=false" >> $GITHUB_OUTPUT
         fi
+      shell: bash
 
     - name: Clear pip cache if pip was repaired
       if: steps.validate_pip.outputs.pip_repaired == 'true'

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -5,6 +5,7 @@ runs:
   steps:
     - name: Merge develop branch into the current branch
       uses: ./.github/actions/merge
+
     - name: Setup Python 3.9.20
       uses: actions/setup-python@v5
       id: setup_python
@@ -14,14 +15,17 @@ runs:
         cache: 'pip'
         cache-dependency-path: |
           requirements_dev.txt
+
     - name: Show whether a cache hit occurred (for debugging purposes).
       run: |
         echo "${{ steps.setup_python.outputs.cache-hit }}"
       shell: bash
+
     - name: Write the exact Python version to a file for cache-busting.
       run: |
         echo "${{ steps.setup_python.outputs.python-version }}" > /home/runner/work/.pyversion
       shell: bash
+
     - name: Cache Python dev dependencies
       id: restore_python_dev_dependencies_cache
       uses: actions/cache@v4
@@ -31,8 +35,38 @@ runs:
         key: ${{ runner.os }}-proddeps-${{ hashFiles('/home/runner/work/.pyversion', 'requirements_dev.txt', 'scripts/install_python_dev_dependencies.py') }}
         path: |
           /opt/hostedtoolcache/Python/3.9.20/x64
+
+    - name: Validate and fix pip installation
+      id: validate_pip
+      run: >
+        echo "Checking pip installation...";
+        # Try to run pip debug to check if pip is working;
+        if ! python -m pip debug; then;
+          echo "Pip appears to be broken, need to fix before proceeding";
+          # Remove pip completely;
+          python -m pip uninstall -y pip setuptools;
+          
+          # Use get-pip.py to get a basic pip installation working;
+          # In Install third-party Python dev dependencies step,;
+          # correct version of pip will be installed;
+          curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py;
+          python get-pip.py;
+          
+          # Flag that we had to repair pip;
+          echo "pip_repaired=true" >> $GITHUB_OUTPUT;
+        else;
+          echo "Pip appears to be working";
+          echo "pip_repaired=false" >> $GITHUB_OUTPUT;
+        fi;
+
+    - name: Clear pip cache if pip was repaired
+      if: steps.validate_pip.outputs.pip_repaired == 'true'
+      run: |
+        echo "Clearing pip cache due to repair"
+        rm -rf ~/.cache/pip
+
     - name: Install third-party Python dev dependencies
-      if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit == false
+      if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit == false || steps.validate_pip.outputs.pip_repaired == 'true'
       uses: nick-fields/retry@v3
       with:
         max_attempts: 3
@@ -42,12 +76,14 @@ runs:
         # Note that the effects of this don't seem to be cached by setup-python,
         # resulting in the occasional "No module named PIL" errors in CI.
         command: python -m scripts.install_python_dev_dependencies
+
     - name: Show where pip installs its packages, and the contents of that folder.
       run: |
         pip show certifi
         ls -la /opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages
         ls -la /opt/hostedtoolcache/Python/3.9.20/x64/bin
       shell: bash
+
     - name: Check Yarn Cache
       run: |
         # Move outside repository root to ignore the configs in
@@ -64,6 +100,7 @@ runs:
         # Move back to the repository root.
         cd -
       shell: bash
+
     # Production pip dependencies are installed in a custom folder rather than
     # the pip cache (see "Caching Python dependencies" in the README at
     # https://github.com/actions/setup-python?tab=readme-ov-file).
@@ -80,6 +117,7 @@ runs:
       run: |
         echo "${{ steps.setup_python.outputs.python-version }}" > /home/runner/work/.pyversion
       shell: bash
+
     - name: Cache third_party and node_modules
       id: restore_third_party_and_node_modules_cache
       uses: actions/cache@v4
@@ -98,6 +136,7 @@ runs:
           /home/runner/work/oppia/oppia/node_modules
           /home/runner/work/oppia/oppia/third_party/python_libs
           /home/runner/work/oppia/oppia/third_party/static
+
     - name: Install third-party libs on cache miss (retries on error)
       if: steps.restore_third_party_and_node_modules_cache.outputs.cache-hit == false
       uses: nick-fields/retry@v3
@@ -107,6 +146,7 @@ runs:
         shell: bash
         timeout_minutes: 30
         command: python -m scripts.install_third_party_libs
+
     - name: Check Yarn Cache
       # Check the yarn cache again to see what packages were just
       # installed.

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -65,6 +65,7 @@ runs:
       run: |
         echo "Clearing pip cache due to repair"
         rm -rf ~/.cache/pip
+      shell: true
 
     - name: Install third-party Python dev dependencies
       if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit == false || steps.validate_pip.outputs.pip_repaired == 'true'

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -37,7 +37,7 @@ runs:
           /opt/hostedtoolcache/Python/3.9.20/x64
 
     # TODO (#22567): Fix pip not working properly when fetched from cache.
-    - name: Check pip installation and fix if needed
+    - name: Check cached pip installation and fix if needed
       run: |
         set -e
         echo "Checking if pip is properly installed..."

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -37,27 +37,27 @@ runs:
           /opt/hostedtoolcache/Python/3.9.20/x64
 
     - name: Validate and fix pip installation
-      id: validate_pip
-      run: >
-        echo "Checking pip installation...";
-        # Try to run pip debug to check if pip is working;
-        if ! python -m pip debug; then;
-          echo "Pip appears to be broken, need to fix before proceeding";
-          # Remove pip completely;
-          python -m pip uninstall -y pip setuptools;
-          
-          # Use get-pip.py to get a basic pip installation working;
-          # In Install third-party Python dev dependencies step,;
-          # correct version of pip will be installed;
-          curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py;
-          python get-pip.py;
-          
-          # Flag that we had to repair pip;
-          echo "pip_repaired=true" >> $GITHUB_OUTPUT;
-        else;
-          echo "Pip appears to be working";
-          echo "pip_repaired=false" >> $GITHUB_OUTPUT;
-        fi;
+      id: validate_pip          
+      run: |
+        echo "Checking pip installation..."
+        # Try to run pip debug to check if pip is working
+        if ! python -m pip debug; then
+          echo "Pip appears to be broken, need to fix before proceeding"
+          # Remove pip completely
+          python -m pip uninstall -y pip setuptools
+
+          # Use get-pip.py to get a basic pip installation working
+          # In Install third-party Python dev dependencies step,
+          # correct version of pip will be installed
+          curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py
+          python get-pip.py
+
+          # Flag that we had to repair pip
+          echo "pip_repaired=true" >> $GITHUB_OUTPUT
+        else
+          echo "Pip appears to be working"
+          echo "pip_repaired=false" >> $GITHUB_OUTPUT
+        fi
 
     - name: Clear pip cache if pip was repaired
       if: steps.validate_pip.outputs.pip_repaired == 'true'

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -65,7 +65,7 @@ runs:
       run: |
         echo "Clearing pip cache due to repair"
         rm -rf ~/.cache/pip
-      shell: true
+      shell: bash
 
     - name: Install third-party Python dev dependencies
       if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit == false || steps.validate_pip.outputs.pip_repaired == 'true'

